prot.c

```diff
@@ -1,15 +1,29 @@
+#include "prot.h"
 #include <sys/types.h>
-#include <unistd.h>
 #include <grp.h>
-#include "prot.h"
+#include <unistd.h>
 
-int prot_gid(gid_t gid)
+#ifndef GROUP_COUNT
+/* use a sane default, but one can override it from outside if there are
+ * higher requirements. */
+#define GROUP_COUNT 8
+#endif
+int prot_gids(const char *user, gid_t gid)
 {
-  if (setgroups(1,&gid) == -1) return -1;
-  return setgid(gid); /* _should_ be redundant, but on some systems it isn't */
-}
+  gid_t gids[GROUP_COUNT];
+  int gcount = GROUP_COUNT;
 
-int prot_uid(uid_t uid)
-{
-  return setuid(uid);
+  if (user) {
+    int r = getgrouplist(user, gid, gids, &gcount);
+    /* member of too many groups */
+    if (r < 0) {
+      return r;
+    }
+  } else {
+    gcount = 1;
+    gids[0] = gid;
+  }
+
+  if (setgroups(gcount,gids) == -1) return -1;
+  return setgid(gid); /* _should_ be redundant, but on some systems it isn't */
```

Comment:
We should always setgid, as setgroups only deals with secondary groups (this also makes the else-clause above useless).

Reply:
The code only avoids the call to setgid() if setgroups() fails.

Reply:
I misread this. Not a fan of return on the same line. ;)